### ISSUE TITLE
Overview of building .tex files for the examples

### ISSUE THREAD

**Issue by niazim3** [JacquesCarette/Drasil#701]
When building the latex documents for the examples, the following problems come up:

- images not found (which makes sense as only GlassBR has a filepath that allows access to images)
- undefined control sequence for 
    - `\tableofcontents`  (all examples)
    - `^{\circ}$C & temperature (centigrade)`
- `Missing } inserted.` messages
- `Missing \endgroup inserted.`
- overall inability to generate a PDF (SSP, SWHS)

Issues should be made to work on avoiding these messages, errors and warnings.

Details below:
***
## Chipmunk, GlassBR:
```
! Undefined control sequence.
<recently read> \

l.25 \tableofcontents
```
```
Underfull \vbox (badness 10000) detected at line 50
! Undefined control sequence.
<recently read> \

l.195 \end{longtabu}
```
[Chipmunk_SRS.pdf](https://github.com/JacquesCarette/Drasil/files/2128818/Chipmunk_SRS.pdf)
is the resulting file, which has an empty Contents section, an empty References section, and ?? links to the referenced sections and contents.

***
## NoPCM
(above warnings) and
```
! Undefined control sequence.
<recently read> \

l.45 ${
     }^{\circ}$C & temperature (centigrade)
```
```
LaTeX Warning: Citation 'parnasClements1986' on page 4 undefined on input line
182.

[4]

LaTeX Warning: Reference `Sec:UserChars' on page 5 undefined on input line 188.



LaTeX Warning: Reference `Sec:References' on page 5 undefined on input line 191
.


LaTeX Warning: Reference `Sec:IMs' on page 5 undefined on input line 191.


LaTeX Warning: Reference `Sec:IMs' on page 5 undefined on input line 192.


LaTeX Warning: Reference `Figure:SysCon' on page 5 undefined on input line 198.



LaTeX Warning: File `SystemContextFigure.png' not found on input line 201.


! Package luatex.def Error: File `SystemContextFigure.png' not found.

See the luatex.def package documentation for explanation.
Type  H <return>  for immediate help.
 ...

l.201 ...[width=\textwidth]{SystemContextFigure.png}

```
```
(load luc: C:/Users/Owner/AppData/Local/MiKTeX/2.9/luatex-cache/generic/fonts/o
tl/lmmono12-regular.luc)

LaTeX Warning: Reference `Figure:SysCon' on page 5 undefined on input line 202.



LaTeX Warning: Reference `Figure:SysCon' on page 5 undefined on input line 202.


[5] [6]

LaTeX Warning: Reference `Figure:Tank' on page 7 undefined on input line 243.

[7]

LaTeX Warning: File `TankWaterOnly.png' not found on input line 250.


! Package luatex.def Error: File `TankWaterOnly.png' not found.

See the luatex.def package documentation for explanation.
Type  H <return>  for immediate help.
 ...

l.250 ...aphics[width=\textwidth]{TankWaterOnly.png}

```
Multiple: 
```
! Missing \endgroup inserted.
<inserted text>
\endgroup
l.423 \end{dmath}

```
```
! Package luatex.def Error: File `ATrace.png' not found.

See the luatex.def package documentation for explanation.
Type  H <return>  for immediate help.
 ...

l.640 \includegraphics[width=\textwidth]{ATrace.png}

```
```
! Package luatex.def Error: File `RTrace.png' not found.

See the luatex.def package documentation for explanation.
Type  H <return>  for immediate help.
 ...

l.647 \includegraphics[width=\textwidth]{RTrace.png}

```
[NoPCM_SRS.pdf](https://github.com/JacquesCarette/Drasil/files/2128847/NoPCM_SRS.pdf)
***
## SWHS 
(similar to NoPCM's but) also:
```
LaTeX Warning: Hyper reference `DD:ht.flux.P' on page 22 undefined on input lin
e 746.

[22]
! Extra }, or forgotten \endgroup.
l.748 ...\left({T_{C}}-{T_{W}}\left(t\right)\right)}
                                                  \,dt-\int_{0}^{t}{{h_{P}} ...
```  
Ends at 
```
! Missing } inserted.
<inserted text>
}
l.749 \end{dmath}

```
and produces a damaged file that cannot be opened.

***

## SSP 
(similar to GlassBR/GamePhysics) with additional:
```
! Undefined control sequence.
<recently read> \

l.45 ${
     }^{\circ}$ & angle (degree)
?
```
```
! Package luatex.def Error: File `IndexConvention.png' not found.

See the luatex.def package documentation for explanation.
Type  H <return>  for immediate help.
 ...

l.330 ...hics[width=\textwidth]{IndexConvention.png}

```
```

See the luatex.def package documentation for explanation.
Type  H <return>  for immediate help.
 ...

l.337 ...raphics[width=\textwidth]{ForceDiagram.png}
```
```

[19] [20] [21] [22] [23] [24]
! Missing \endgroup inserted.
<inserted text>
\endgroup
l.851 \end{dmath}

```
Ends at 
```
! Missing } inserted.
<inserted text>
}
l.1122 \end{dmath}
```
and doesn't produce an openable file just like SWHS.

***
## Tiny
The table of contents message already mentioned, and 
```
[1{C:/ProgramData/MiKTeX/2.9/pdftex/config/pdftex.map}]
! Undefined control sequence.
<recently read> \

l.34 ${
     }^{\circ}$C & temperature (centigrade)
?
```

**Comment by JacquesCarette** [comment]
Is this with `lualatex`? I know that things go awry with `pdflatex` (i.e. the \tableofcontents error seems familiar). In fact, we should probably make sure that `pfdlatex` doesn't even try.

Quite a few could be related to that, as well as to unicode and/or line endings.

**Comment by niazim3** [comment]
Yes, this is with `lualatex`. The makefiles all have the following commands built in (where GlassBR_SRS is replaced by the actual example file names):
```
lualatex GlassBR_SRS
-bibtex GlassBR_SRS
lualatex GlassBR_SRS
lualatex GlassBR_SRS
```

When I run it on my machine, it says I have the following version by the way, which seems to be the most recent release `This is LuaTeX, Version 1.0.4 (MiKTeX 2.9.6300) restricted system commands enabled.`

**Comment by JacquesCarette** [comment]
Hmm, googling around, it seems that adding option `unicode` to `hyperref` might help with the first part.

**Comment by JacquesCarette** [comment]
`\circ` is in package `amssymb` -- is it loaded?

**Comment by niazim3** [comment]
Yes, it is loaded 
https://github.com/JacquesCarette/Drasil/blob/628b54d0e17d0b818be0c36480416e5bf37a2139/code/stable/nopcm/NoPCM_SRS.tex#L7

**Comment by niazim3** [comment]
Sorry, @JacquesCarette adding `\usepackage[unicode=true]{hyperref}` will contribute to which part exactly? I searched around online and it doesn't seem related to the items mentioned in the original overview...

**Comment by JacquesCarette** [comment]
I think it's just `usepackage[unicode]{hyperref}`. To the `\tableofcontents` issue - if the table of contents contains unicode characters, then processing it will fail. And might give uninformative errors.

**Comment by JacquesCarette** [comment]
But this is all just wild guesses on my part too. I've encountered similar things before, but forgot what I did to fix them.  Eventually I think I had to use bisection -- keep chopping the file in 'half' (while keeping the relevant stuff) to narrow down the smallest reproducible case.

Generally, I comment out the whole bottom of the file below the error, and then slowly comment out things above it, until I get a very small file that exhibits the problem. And then comment out 1 additional line that makes the error go away, which then tells me the problem is definitely related to that line.

**Comment by JacquesCarette** [comment]
I wouldn't create sub-issues -- it may well be that all the problems are related!

**Comment by smiths** [comment]
@niazim3, thank you for taking on this issue.  We tend to worry about the html files, since they are easier to check, but, we want the tex files to work too.  To be honest, I much prefer the look of the tex files.  Jacques advice for debugging is good.  I do it the same way with tex files, especially when the error messages don't make any sense.  😄

**Comment by niazim3** [comment]
Just noting here, to further look into why the placement of `\maketitle` is influencing the presence of the `tableofcontents` error.

**Comment by niazim3** [comment]
It seems that the `\tableofcontents` issue is related to what's imported. After searching online and testing minimal text (solutions to which are documented in the previous comment), the following link seems most useful in terms of a potential recognition of how this can be fixed: [amsmath + breqn + unicode-math breaks](http://tug.org/pipermail/lualatex-dev/2011-August/001290.html).

**Comment by JacquesCarette** [comment]
Do we use any of "breqn, mathstyle and flexisym"?

**Comment by niazim3** [comment]
Breqn was added to deal with "Formatting Long Equations in TeX" (i.e. #346)

https://github.com/JacquesCarette/Drasil/blob/628b54d0e17d0b818be0c36480416e5bf37a2139/code/Language/Drasil/TeX/Preamble.hs#L23

As for mathstyle and flexisym, they are not used in Drasil.

**Comment by JacquesCarette** [comment]
Thanks.  What happens if you take `breqns` out, does it work?

**Comment by niazim3** [comment]
I should specify that by running the `lualatex _` command twice manually, the table of contents section does appear for the examples it originally wasn't appearing for (i.e. GlassBR, GamePhysics, NoPCM) just by running `make`; although this workaround isn't working with the damaged examples (i.e. SSP, SWHS). 
***
#### Re: By taking `breqns` out completely:

For the non-damaged examples (such as GlassBR), by removing/commenting out use of `breqns`, running `make` follows through in the way it's supposed to; that is, table of contents and references appear (probably from actually running the commands in the makefile, whereas before it seemed that `lualatex` was the only command being run).

For SSP, running make now works and produces a non-damaged file. However, it causes progress made by #346 to be undone, as the equations are going off the screens.

**Comment by JacquesCarette** [comment]
Hmm, I think that lualatex (and unicode) is a more important feature than automatic linebreaking, which is always very fragile.  We should open a new issue with just the long equations, and otherwise accept this change.  And remove `breqns` from our list of required packages.

**Comment by niazim3** [comment]
I'll make the changes in a separate branch.

Should a new issue be opened or should the original "Formatting Long Equations in TeX #346" be reopened?

**Comment by JacquesCarette** [comment]
New issue. #346 had many things in it, most of which were indeed fixed.

**Comment by niazim3** [comment]
Removal of `breqn` has led to multiple warnings to pop up in relation to the following:
```
! LaTeX Error: Environment dmath undefined.
```

. As mentioned earlier, PDFs are still produced; but I didn't realize that the `dmath` environment was related to the `breqn` package when I removed it.

It seems that Drasil recently migrated to using `dmath` instead of `$$` (which, after searching online, apparently is a deprecated way of using math environments ([see what-are-the-differences-between-align-equation-and-displaymath](https://tex.stackexchange.com/questions/40492/what-are-the-differences-between-align-equation-and-displaymath))) in the [Devi](https://github.com/JacquesCarette/Drasil/pull/530) branch: 
![image](https://user-images.githubusercontent.com/28247301/42651831-9758a5a8-85de-11e8-824d-6d0a2c49d11c.png)
and the switch from `equation` to `dmath` was done in commit ca5a4c3. Should that `equation` be brought back even though equations are output (despite the warnings) right now? @JacquesCarette @szymczdm

**Comment by JacquesCarette** [comment]
We should be using `\[` and `\]` rather than `$$`. [And I read that it should be `\(` and `\)` rather than `$` as well]. That's probably worth making into a new issue.  And be 'fixed' once (at least) the yuzhi branch is merged in, not before.

**Comment by JacquesCarette** [comment]
The original issue is fixed.
